feat(aws-sqs): Add external ID support for cross-account access

tangobango5 opened the pull request:
This commit adds support for external ID in AWS SQS scaler to enable secure cross-account access scenarios. External ID is now parsed from TriggerAuthentication and passed to STS AssumeRole operations.

Changes:
- Add AwsExternalID field to AuthorizationMetadata struct
- Update GetAwsAuthorization to parse external ID from auth parameters
- Modify cache key generation to include external ID
- Update AssumeRole providers to use external ID when available
- Add comprehensive test coverage for external ID scenarios

The external ID is only used with AssumeRole operations and maintains backward compatibility with existing configurations.

<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md
-->

_Provide a description of what has been changed_

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

<!--
  Make sure to link the related issue for this change
  If it requires multiple PRs and/or a PR on another repo as well, please use "Relates to" instead.
-->
Fixes # https://github.com/kedacore/keda/issues/6921

<!--
  Make sure to link the related PRs for changes such as documentation & Helm charts
-->
Relates to #

tangobango5 committed:
feat(aws-sqs): Add external ID support for cross-account access

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 committed:
Add custom KEDA chart with external ID support

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

JorTurFer reviewed (commented):
Awesome improvement! ❤️ 
As this is part of pod identity (IRSA role assumption), WDYT if we set it as a new parameter in trigger authentication? This would make the new feature available for any AWS Scaler at once.

Please, also open a PR to docs to document the new parameter

zroubalik commented:
> Awesome improvement! ❤️ As this is part of pod identity (IRSA role assumption), WDYT if we set it as a new parameter in trigger authentication? This would make the new feature available for any AWS Scaler at once.
> 

Could you please elaborate here?

JorTurFer commented:
> > Awesome improvement! ❤️ As this is part of pod identity (IRSA role assumption), WDYT if we set it as a new parameter in trigger authentication? This would make the new feature available for any AWS Scaler at once.
> 
> Could you please elaborate here?

sure, `AwsExternalID` is part of the auth SDK and not from SQS SDK. The OP has added the parameter to the SQS trigger but it's passed directly to the auth SDK. The auth code is shared for all the AWS Scalers, so I guess that it's a potential improvement for all the AWS Scalers and not only for SQS. If any other user needs `AwsExternalID` for other AWS Scaler, we will need to add the parameter there and pass to the shared auth code. 
My suggestion is to remove the parameter from the SQS trigger and moving it directly to the TriggerAuthentication to make in shared for any AWS scaler and also to place it closer to the auth code and not in the SQS code to pass it till auth code. The last but not least, this change ONLY applies if pod identity is used (the assumeRole code is podIdentity code), so I see it configured along podIdentity configuration and not as part of the scaler

tangobango5 committed:
Use podIdentity

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 committed:
Fix changes

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

JorTurFer reviewed (commented):
awesome job! Just one extra thing, instead of using SQS unit test to test this feature, what about adding an e2e test for it. We have a real AWS account where we can include a role with externalID requirement -> https://github.com/[kedacore/testing-infrastructure](https://github.com/kedacore/testing-infrastructure)

The AWS IAM is managed here -> https://github.com/kedacore/testing-infrastructure/blob/main/terraform/modules/aws/iam/main.tf

tangobango5 committed:
test: add e2e test for AWS external ID support in pod identity

Add end-to-end test to verify the externalID field in TriggerAuthentication
spec works correctly with AWS pod identity authentication.

The test covers:
- Scaling with correct external ID (should scale successfully)
- Scaling without external ID when role requires it (should fail to scale)

Required environment variables:
- TF_AWS_ROLE_ARN_EXTERNAL_ID: ARN of IAM role requiring external ID
- TF_AWS_EXTERNAL_ID: The external ID configured in role's trust policy

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 committed:
fix: use context.Background() instead of context.TODO() in test

Fixes Semgrep lint warning about using well-defined context.

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 committed:
Merge main

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 committed:
Fix crd

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 committed:
Fix changelog
Add externalID tests

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 commented:
@JorTurFer you may review now.

JorTurFer reviewed (commented):
LGTM! Only one small thing: please include the externalID in the key generated by `getCacheKey`. I assume that you can never assume a role with and without the externalD, but including the value in the key ensures that connections are unique. Also, if one legit trigger is configured, not having the externalID can allow another one wrongly configured (without it) to use the right one

rickbrouwer commented:
Related with #7665 and #7580

tangobango5 committed:
Merge branch 'main' into feature/aws-sqs-external-id-support

Signed-off-by: Ritesh Kumar Chaurasia <riteshchaurasia53@gmail.com>

tangobango5 committed:
Fix changelog order

Signed-off-by: ritesh.chaurasia1 <ritesh.chaurasia@devo.com>

tangobango5 commented:
@JorTurFer @rickbrouwer 
I think all comments are addressed.
Please do let me know if further changes are required here.

tangobango5 committed:
Merge branch 'main' into feature/aws-sqs-external-id-support

rickbrouwer commented:
/run-e2e aws
**Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/25206115443)

zeevimeytar commented:
Can we expect this to be part of the v2.20 release?
